Unify Delta Lake tests' container handling with Hive and Iceberg

mdesmet opened the pull request:
<!-- Thank you for submitting a pull request! Find more information in our development guide at https://github.com/trinodb/trino/blob/master/.github/DEVELOPMENT.md and contact us on #dev in Slack. -->

## Description

The `MinioClient` provides  additional features on top of the standard S3 api that may be interesting to keep in our toolbox. So it has been integrated the `HiveMinioDataLake` container setup:

- Uploading of files based on resources (similar to below missing HadoopHive feature)
- Event notifications on when files are created or updated similar to AWS lambda (not currently used though)

`BaseTestContainer` has been extended for executing commands in a container and retrieving the result and loading resources.

> Is this change a fix, improvement, new feature, refactoring, or other?

> Is this a change to the core query engine, a connector, client library, or the SPI interfaces? (be specific)

Change only impacting tests for Hive, Iceberg and Delta Lake and common test libraries.

> How would you describe this change to a non-technical end user or system administrator?

N/A

## Related issues, pull requests, and links

* Fixes #11300

<!-- The following sections are filled in by the maintainer with input from the contributor:
Use :white_check_mark: or (x) to signal selection.
-->

## Documentation

(x) No documentation is needed.

## Release notes

(x) No release notes entries required.

ebyhr commented on plugin/trino-delta-lake/src/test/java/io/trino/plugin/deltalake/TestDeltaLakeAdlsConnectorSmokeTest.java:
```diff
@@ -19,38 +19,25 @@
 import com.google.common.reflect.ClassPath;
-import io.trino.plugin.deltalake.util.DockerizedDataLake;
-import io.trino.plugin.deltalake.util.TestingHadoop;
+import io.trino.plugin.hive.containers.HiveHadoop;
```
Please implement both `createHiveMinioDataLake` & `createDeltaLakeQueryRunner` methods.

mdesmet replied:
I've implemented those methods, I could see the hive container launching now. Let me know if it works on your side.

ebyhr replied:
I'm facing the below exception now:
```
java.lang.RuntimeException: Got exception: java.io.IOException No FileSystem for scheme: abfs
```

mdesmet replied:
I missed the overridden Hadoop version. Can you retest? Don't have an Azure account at the moment.

ebyhr replied:
It works now, thanks!

ebyhr commented on plugin/trino-delta-lake/src/test/java/io/trino/plugin/deltalake/TestDeltaLakeAdlsStorage.java:
```diff
@@ -16,10 +16,10 @@
 import com.google.common.io.Resources;
-import io.trino.plugin.deltalake.util.DockerizedDataLake;
-import io.trino.plugin.deltalake.util.TestingHadoop;
+import io.trino.plugin.hive.containers.HiveHadoop;
```
This test class looks broken now. I can't run locally. Please note that CI doesn't run this test because of forked branch PR.
```
java.lang.IllegalArgumentException: Resource with path /var/folders/8s/dkvf18z55lj_9yxhy1n54sph0000gn/T/core-site16451568450651754371.xml could not be found on any of these classloaders: [jdk.internal.loader.ClassLoaders$AppClassLoader@3b192d32]
```

mdesmet replied:
I've tested the class locally by removing the constructor. The container is launched successfully now and I can also execute hadoop fs commands. Let me know if it works on your side.

ebyhr replied:
It works now, thanks!

ebyhr commented on plugin/trino-delta-lake/src/test/java/io/trino/plugin/deltalake/TestDeltaLakeAnalyze.java:
```diff
@@ -16,7 +16,7 @@
 import io.trino.Session;
 import io.trino.operator.OperatorStats;
-import io.trino.plugin.deltalake.util.DockerizedMinioDataLake;
+import io.trino.plugin.hive.containers.HiveMinioDataLake;
```
This test class looks broken now. I can't run locally after removing `abstract`. 
```
java.lang.IllegalStateException: Mapped port can only be obtained after the container is started
```

mdesmet replied:
it was missing `hiveMinioDataLake.start()`. Succesfully tested locally.

ebyhr replied:
It works now, thanks!

findinpath commented:
There would be a few follow-ups (very likely a different PR to avoid this one exploding):

- Extract the `HiveMinioDataLake` and the other classes to `trino-testing-containers`.
- Adapt `trino-faulttolerant-tests` to drop the dependency and use the infrastructure previously moved to `trino-testing-containers` instead

https://github.com/trinodb/trino/blob/6b707525f356ffdbad23188a92c5aba033641405/testing/trino-faulttolerant-tests/pom.xml#L148-L153

findinpath commented:
Please rebase the PR.

findinpath commented on plugin/trino-hive/src/test/java/io/trino/plugin/hive/containers/HiveMinioDataLake.java:
```diff
@@ -89,6 +93,7 @@ public void start()
                 .build();
         s3Client.createBucket(this.bucketName);
```
Why do we have both s3client and minioclient?
I think that the s3client is now superfluos.

aczajkowski replied:
👍  for that. One or another. My initial idea was to use S3Client as this would allow to have some base class in future which would handle real S3 tests and Minio based test.
For now it seems we should stick to one selected client.

findinpath commented on plugin/trino-delta-lake/src/test/java/io/trino/plugin/deltalake/BaseDeltaLakeConnectorSmokeTest.java:
```diff
@@ -118,11 +117,26 @@ abstract QueryRunner createDeltaLakeQueryRunner(Map<String, String> connectorPro
 
+    protected void copyResources(String resourcePath, String target)
+    {
+        hiveMinioDataLake.getMinioClient().copyResourcePath(bucketName, resourcePath, target);
```
`hiveMinioDataLake.getMinioClient().copyResourcePath` keeps appearing multiple times in the PR.

Would it make sense to expose in `HiveMinioDataLake` (same as it was done in `DockerizedMinioDataLake` the methods  :

- copyResources
- writeFile
- listFiles

?

findinpath commented on plugin/trino-delta-lake/src/test/java/io/trino/plugin/deltalake/BaseDeltaLakeAwsConnectorSmokeTest.java:
```diff
@@ -26,23 +25,20 @@
-                ImmutableMap.of("io/trino/plugin/deltalake/core-site.xml", "/etc/hadoop/conf/core-site.xml"),
-                ImmutableMap.of());
-        return dockerizedMinioDataLake;
+        hiveMinioDataLake = new HiveMinioDataLake(bucketName, ImmutableMap.of(), "ghcr.io/trinodb/testing/hdp2.6-hive");
```
Pls avoid using magic strings for the docker image - Let's introduce a constant for this purpose.

aczajkowski replied:
NIT: I remember there is a class where we store docker image version. 
`io.trino.plugin.deltalake.util.DockerImages`
Maybe let's extract it to test containers and keep those values there

findinpath commented on plugin/trino-delta-lake/src/test/java/io/trino/plugin/deltalake/TestDeltaLakeAdlsConnectorSmokeTest.java:
```diff
@@ -89,30 +88,26 @@ DockerizedDataLake createDockerizedDataLake()
+        HiveMinioDataLake hiveMinioDataLake = new HiveMinioDataLake(
+                bucketName,
+                ImmutableMap.of(hadoopCoreSiteXmlTempFile.normalize().toAbsolutePath().toString(), "/etc/hadoop/conf/core-site.xml"),
+                "ghcr.io/trinodb/testing/hdp3.1-hive");
```
Pls avoid using magic strings for the docker image - Let's introduce a constant for this purpose.

findinpath commented on plugin/trino-delta-lake/src/test/java/io/trino/plugin/deltalake/util/TestingHadoop.java:
```diff
@@ -1,271 +0,0 @@
-        container = Failsafe.with(retryPolicy).get(() -> startHadoopContainer(image, ports, resourcesToMount, filesToMount, extraHosts, network));
-        closer.register(container::stop);
-
-        // HACK: even though Hadoop is accessed by proxy, Hadoop still tries to resolve hadoop-master
```
@losipiuk I just want to make sure that this logic is not needed with the transition towards `HiveHadoop` where this setting is not made anymore.

aczajkowski commented on testing/trino-testing-containers/src/main/java/io/trino/testing/containers/BaseTestContainer.java:
```diff
@@ -124,6 +132,38 @@ public void stop()
+            }
+        }
+        catch (IOException | InterruptedException e) {
+            throw new RuntimeException("Exception while running command: " + printableCommand(commandAndArgs), e);
```
NIT: Just use String.join
```
String.join(" ", commandAndArgs)
```

aczajkowski commented on testing/trino-testing-containers/src/main/java/io/trino/testing/containers/BaseTestContainer.java:
```diff
@@ -124,6 +132,38 @@ public void stop()
+    public void runCommandInContainer(String... commandAndArgs)
+    {
+        try {
+            Container.ExecResult execResult = container.execInContainer(commandAndArgs);
```
NIT: call public method from this class to not have redundant code.
```
Container.ExecResult execResult = executeInContainer(commandAndArgs)
```

aczajkowski commented on testing/trino-testing-containers/src/main/java/io/trino/testing/containers/BaseTestContainer.java:
```diff
@@ -76,7 +83,8 @@ protected void setupContainer()
             container.addExposedPort(port);
         }
-        filesToMount.forEach(this::copyFileToContainer);
+        resourcesToMount.forEach(this::copyFileToContainer);
```
NIT: If we made that change so let's maybe rename copyFileToContainer to copyResourceToContainer

aczajkowski commented on plugin/trino-hive/src/test/java/io/trino/plugin/hive/containers/HiveMinioDataLake.java:
```diff
@@ -61,15 +62,18 @@ public HiveMinioDataLake(String bucketName, Map<String, String> hiveHadoopFilesT
+        HiveHadoop.Builder hiveHadoopBuilder = HiveHadoop.builder()
+                .withImage(hiveHadoopImage)
+                .withNetwork(network);
+        if (hiveHadoopFilesToMount.isEmpty()) {
```
Not sure here but shouldn't we always copy resources "hive_minio_datalake/hive-core-site.xml"
and if there are any files to mount then also files like
```
HiveHadoop.Builder hiveHadoopBuilder = HiveHadoop.builder()
                .withImage(hiveHadoopImage)
                .withNetwork(network)
                .withResourcesToMount(ImmutableMap.of("hive_minio_datalake/hive-core-site.xml", "/etc/hadoop/conf/core-site.xml"));
if (!hiveHadoopFilesToMount.isEmpty()) {
    hiveHadoopBuilder.withFilesToMount(hiveHadoopFilesToMount);
}
```
We could potentially just blindly allways add withFilesToMount. It it's empty forEach will not iterate and do not mount them

aczajkowski commented on plugin/trino-iceberg/pom.xml:
```diff
@@ -355,6 +355,29 @@
             <scope>test</scope>
         </dependency>
 
+        <dependency>
```
👍  for that. But I assume it's a separate PR to bump this version.

aczajkowski commented on plugin/trino-hive/src/test/java/io/trino/plugin/hive/containers/MinioClient.java:
```diff
@@ -45,7 +45,7 @@
 import static java.util.regex.Matcher.quoteReplacement;
 
-class MinioClient
+public class MinioClient
```
NIT: When we use MinioClient now directly in container. I'm wondering if we could maybe move client creation to MinIo container and fetch it from them anywhere it's needed ?

aczajkowski commented on plugin/trino-hive/src/test/resources/hive_minio_datalake/hive-core-site.xml:
```diff
@@ -24,4 +24,15 @@
     </property>
+
+    <!-- Hive impersonation -->
+    <property>
```
Why do we need them now. I can't see changes that might require it ?

aczajkowski replied:
Ok I found it in next commit :D

aczajkowski commented on plugin/trino-delta-lake/src/test/java/io/trino/plugin/deltalake/TestDeltaLakeAdlsConnectorSmokeTest.java:
```diff
@@ -89,30 +88,26 @@ DockerizedDataLake createDockerizedDataLake()
     {
-        if (adlsDirectory != null && dockerizedDataLake.getTestingHadoop() != null) {
-            dockerizedDataLake.getTestingHadoop().runCommandInContainer("hadoop", "fs", "-rm", "-f", "-r", adlsDirectory);
+        if (adlsDirectory != null && hiveMinioDataLake.getHiveHadoop() != null) {
```
i think `hiveMinioDataLake.getHiveHadoop() != null` is not needed. It will never be null when using HibeMinioDataLake

aczajkowski reviewed (commented):
Nice % few nit comments

aczajkowski reviewed (approved):
LGTM

findinpath commented on plugin/trino-delta-lake/src/test/java/io/trino/plugin/deltalake/TestDeltaLakeSharedHiveMetastoreWithViews.java:
```diff
@@ -94,30 +92,30 @@ public void testReadInformationSchema()
         assertQuery("SHOW TABLES FROM hive." + schema, "VALUES 'hive_table', 'hive_view', 'delta_table'");
 
-        assertThatThrownBy(() -> query("SHOW CREATE TABLE delta." + schema + ".hive_table"))
+        assertThatThrownBy(() -> query("SHOW CREATE TABLE delta_lake." + schema + ".hive_table"))
```
optional: Using `io.trino.plugin.deltalake.DeltaLakeQueryRunner#DELTA_CATALOG` would be also on option.

mdesmet replied:
We could even omit `delta_lake` but thought this way is a bit more explicit.

aczajkowski commented on plugin/trino-hive/src/test/java/io/trino/plugin/hive/containers/HiveMinioDataLake.java:
```diff
@@ -44,6 +59,11 @@
     private State state = State.INITIAL;
     private AmazonS3 s3Client;
```
+1 for that. I would stick to one or another. Let's not maintain both.

findinpath commented on plugin/trino-hive/src/test/java/io/trino/plugin/hive/containers/HiveMinioDataLake.java:
```diff
@@ -61,15 +75,15 @@ public HiveMinioDataLake(String bucketName, Map<String, String> hiveHadoopFilesT
+        HiveHadoop.Builder hiveHadoopBuilder = HiveHadoop.builder()
+                .withImage(hiveHadoopImage)
+                .withNetwork(network);
+        Map<String, String> defaultHiveHadoopResourcesToMount = new HashMap<>(Map.of(
```
```suggestion
                Map<String, String> defaultHiveHadoopResourcesToMount = ImmutableMap.<String, String>builder()
                .put("/etc/hadoop/conf/core-site.xml", getPathFromClassPathResource("hive_minio_datalake/hive-core-site.xml"))
                .putAll(hiveHadoopFilesToMount)
                .buildOrThrow();
```

mdesmet replied:
I had also tried that, however Guava's `ImmutableMap` doesn't allow duplicate keys to be inserted. Here we sometimes want to override the core-site.xml with a new one. Hence the standard `HashMap`.

findinpath commented on plugin/trino-delta-lake/src/test/java/io/trino/plugin/deltalake/TestDeltaLakePerTransactionMetastoreCache.java:
```diff
@@ -65,16 +64,16 @@
     {
-        if (dockerizedMinioDataLake != null) {
-            dockerizedMinioDataLake.close();
+        if (hiveMinioDataLake != null) {
```
Gentle reminder for handling this comment.

findinpath reviewed (commented):
LGTM % comments

@ebyhr  CPTAL?

findinpath commented:
@mdesmet 

> Remaned ACCESS_KEY and SECRET_KEY to MINIO_ACCESS_KEY and `MINIO_SECRET_KEY` for readability

Use a shorter commit message header and explain the detail in the commit message content.

e.g. _Renamed MinIO credentials constants_

ebyhr commented on plugin/trino-delta-lake/src/test/java/io/trino/plugin/deltalake/DeltaLakeQueryRunner.java:
```diff
@@ -52,6 +53,69 @@
+
+        protected Builder()
+        {
+            super(testSessionBuilder()
+                    .setCatalog(DELTA_CATALOG)
+                    .setSchema(TPCH_SCHEMA)
+                    .build());
```
```suggestion
            super(createSession());
```

ebyhr commented on plugin/trino-hive/src/test/java/io/trino/plugin/hive/containers/HiveMinioDataLake.java:
```diff
@@ -42,7 +51,12 @@
+        this(bucketName, ImmutableMap.of(), HiveHadoop.DEFAULT_IMAGE);
+    }
 
     public HiveMinioDataLake(String bucketName, Map<String, String> hiveHadoopFilesToMount)
```
Remove an unused constructor.

ebyhr commented on plugin/trino-iceberg/pom.xml:
```diff
@@ -348,6 +348,23 @@
+            </exclusions>
+        </dependency>
+
+
```
Remove a redundant empty line (single empty line is enough).

findinpath commented on plugin/trino-delta-lake/src/test/java/io/trino/plugin/deltalake/DeltaLakeQueryRunner.java:
```diff
@@ -52,6 +53,69 @@
+            super(defaultSession);
+        }
+
+        public Builder setDeltaProperties(Map<String, String> deltaProperties)
```
Note that there exists already `io.trino.plugin.hive.HiveQueryRunner.Builder#setHiveProperties`.

ebyhr replied:
I'm sure most connectors except for Hive & Iceberg use `connectorProperties`, but it's just a nit comment. No need to change.

mdesmet commented:
I have extracted the Builder in a separate commit. I also added `catalogName` in the `DeltaLakeQueryRunner.Builder `class as a follow-up on the `delta` vs `delta_lake` name discussion. That way the signature of the `DeltaLakeQueryRunner` factory methods doesn't change too much.

findinpath commented on plugin/trino-delta-lake/src/test/java/io/trino/plugin/deltalake/DeltaLakeQueryRunner.java:
```diff
@@ -52,6 +54,73 @@
+
+                queryRunner.installPlugin(new TestingDeltaLakePlugin());
+                Map<String, String> deltaProperties = new HashMap<>(this.deltaProperties.buildOrThrow());
+                if (!deltaProperties.containsKey("hive.metastore.uri")) {
```
Having business logic in setting up default values for properties is not a  good idea.
In case we'll want to handle Glue here, we'll further complicate the logic. 
I'd advocate that the client classes of `DeltaQueryRunner` should specify these properties themselves.

See for reference: 
- https://github.com/trinodb/trino/blob/74b324cf0089108492619a5176240c2ba2c8f36a/plugin/trino-delta-lake/src/test/java/io/trino/plugin/deltalake/TestDeltaLakeSharedFileMetastoreWithTableRedirections.java#L49-L50
- https://github.com/trinodb/trino/blob/74b324cf0089108492619a5176240c2ba2c8f36a/plugin/trino-delta-lake/src/test/java/io/trino/plugin/deltalake/TestDeltaLakeSharedGlueMetastoreWithTableRedirections.java#L75-L76

mdesmet replied:
I tend to agree with you. Seems like the `IcebergQueryRunner` also does this though. I would still consider putting it in the QueryRunner class itself as duplicating this logic everywhere seems a bit redundant. WDYT of adding a builder method `withFileMetastore`? Seems more explicit and hides the complexity for the clients of the QueryRunner classes.

mdesmet replied:
I would say, let's keep this out of this PR for now.

findinpath replied:
I'm fine with `withFileMetastore` because it abstracts the logic of setting the file hive metastore.

> let's keep this out of this PR for now.
I would argue that this change is related to the current PR.

ebyhr replied:
Follow-up PR is acceptable to me. Going to merger as-is.

findinpath commented on plugin/trino-delta-lake/src/test/java/io/trino/plugin/deltalake/AbstractTestDeltaLakeCreateTableStatistics.java:
```diff
@@ -85,15 +84,16 @@ protected QueryRunner createQueryRunner()
+        hiveMinioDataLake.start();
         ImmutableMap.Builder<String, String> queryRunnerProperties = ImmutableMap.builder();
         queryRunnerProperties.putAll(additionalProperties());
         QueryRunner queryRunner = DeltaLakeQueryRunner.createS3DeltaLakeQueryRunner(
```
Separate commit: Local variable `queryRunner` is redundant.

findinpath commented:
I find it really great that we have a unified way of dealing with minio data lakes on tests. There's only a bit more left to land this awesome test infrastructure change. 

Great job!!

ebyhr commented:
Merged, thanks!
